File: src/skillguard/governance/rbac.py
"""Role-based access control (RBAC) for SkillGuard API.

Provides user authentication and authorization with configurable
roles and permissions for API endpoints.
"""

from __future__ import annotations

from enum import Enum
from datetime import datetime, timezone
from typing import Any

from pydantic import BaseModel, Field


class Role(str, Enum):
    """User roles with different permission levels."""

    ADMIN = "admin"
    ANALYST = "analyst"
    DEVELOPER = "developer"
    VIEWER = "viewer"


class Permission(str, Enum):
    """Granular permissions for API actions."""

    SCAN_SUBMIT = "scan:submit"
    SCAN_READ = "scan:read"
    RULES_READ = "rules:read"
    RULES_WRITE = "rules:write"
    POLICY_READ = "policy:read"
    POLICY_WRITE = "policy:write"
    MONITOR_READ = "monitor:read"
    MONITOR_WRITE = "monitor:write"
    SKILL_READ = "skill:read"
    AUDIT_READ = "audit:read"
    ADMIN_ALL = "admin:*"


# Default role-to-permission mappings
_ROLE_PERMISSIONS: dict[Role, set[Permission]] = {
    Role.ADMIN: {p for p in Permission},
    Role.ANALYST: {
        Permission.SCAN_SUBMIT,
        Permission.SCAN_READ,
        Permission.RULES_READ,
        Permission.RULES_WRITE,
        Permission.POLICY_READ,
        Permission.MONITOR_READ,
        Permission.MONITOR_WRITE,
        Permission.SKILL_READ,
        Permission.AUDIT_READ,
    },
    Role.DEVELOPER: {
        Permission.SCAN_SUBMIT,
        Permission.SCAN_READ,
        Permission.RULES_READ,
        Permission.SKILL_READ,
        Permission.MONITOR_READ,
    },
    Role.VIEWER: {
        Permission.SCAN_READ,
        Permission.RULES_READ,
        Permission.SKILL_READ,
    },
}


class User(BaseModel):
    """A SkillGuard user."""

    id: str
    username: str
    role: Role
    api_key: str | None = None
    created_at: str = Field(default_factory=lambda: datetime.now(timezone.utc).isoformat())
    active: bool = True


class RBACManager:
    """Manages users, roles, and permission checks.

    Provides a simple in-memory RBAC system for the API.
    Can be extended with database persistence.
    """

    def __init__(self) -> None:
        self._users: dict[str, User] = {}
        self._api_keys: dict[str, str] = {}  # api_key -> user_id

    def add_user(self, user: User) -> None:
        """Register a new user."""
        self._users[user.id] = user
        if user.api_key:
            self._api_keys[user.api_key] = user.id

    def remove_user(self, user_id: str) -> bool:
        """Remove a user by ID."""
        user = self._users.pop(user_id, None)
        if user is None:
            return False
        if user.api_key:
            self._api_keys.pop(user.api_key, None)
        return True

    def get_user(self, user_id: str) -> User | None:
        """Get a user by ID."""
        return self._users.get(user_id)

    def get_user_by_api_key(self, api_key: str) -> User | None:
        """Authenticate a user by API key."""
        user_id = self._api_keys.get(api_key)
        if user_id is None:
            return None
        user = self._users.get(user_id)
        if user and user.active:
            return user
        return None

    def check_permission(self, user_id: str, permission: Permission) -> bool:
        """Check if a user has a specific permission."""
        user = self._users.get(user_id)
        if user is None or not user.active:
            return False

        user_permissions = _ROLE_PERMISSIONS.get(user.role, set())

        # Admin has all permissions
        if Permission.ADMIN_ALL in user_permissions:
            return True

        return permission in user_permissions

    def get_user_permissions(self, user_id: str) -> set[Permission]:
        """Get all permissions for a user."""
        user = self._users.get(user_id)
        if user is None or not user.active:
            return set()
        return _ROLE_PERMISSIONS.get(user.role, set()).copy()

    def list_users(self) -> list[User]:
        """List all registered users."""
        return list(self._users.values())

File: go-skillguard/internal/governance/rbac.go
package governance

import (
	"crypto/rand"
	"encoding/hex"
	"sync"
)

// Role defines RBAC roles.
type Role string

const (
	RoleAdmin     Role = "admin"
	RoleAnalyst   Role = "analyst"
	RoleDeveloper Role = "developer"
	RoleViewer    Role = "viewer"
)

// RolePermissions maps roles to their allowed actions.
var RolePermissions = map[Role]map[string]bool{
	RoleAdmin: {
		"scan": true, "rules:read": true, "rules:write": true,
		"policy:read": true, "policy:write": true,
		"audit:read": true, "audit:export": true,
		"community:read": true, "community:write": true,
		"monitor": true, "server": true, "bom": true,
	},
	RoleAnalyst: {
		"scan": true, "rules:read": true,
		"policy:read": true,
		"audit:read": true, "audit:export": true,
		"community:read": true, "community:write": true,
		"monitor": true, "bom": true,
	},
	RoleDeveloper: {
		"scan": true, "rules:read": true,
		"policy:read": true,
		"community:read": true,
		"bom": true,
	},
	RoleViewer: {
		"scan": true, "rules:read": true,
		"policy:read": true,
		"community:read": true,
	},
}

// APIKey represents an API key with an associated role.
type APIKey struct {
	Key  string `json:"key"`
	Role Role   `json:"role"`
	Name string `json:"name"`
}

// RBACManager manages role-based access control.
type RBACManager struct {
	mu   sync.RWMutex
	keys map[string]APIKey // key -> APIKey
}

// NewRBACManager creates a new RBACManager.
func NewRBACManager() *RBACManager {
	return &RBACManager{
		keys: make(map[string]APIKey),
	}
}

// GenerateKey creates a new API key with the given role and name.
func (r *RBACManager) GenerateKey(role Role, name string) APIKey {
	r.mu.Lock()
	defer r.mu.Unlock()

	b := make([]byte, 24)
	rand.Read(b)
	key := "sg_" + hex.EncodeToString(b)

	apiKey := APIKey{Key: key, Role: role, Name: name}
	r.keys[key] = apiKey
	return apiKey
}

// ValidateKey checks if an API key is valid and returns the associated role.
func (r *RBACManager) ValidateKey(key string) (Role, bool) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	ak, ok := r.keys[key]
	if !ok {
		return "", false
	}
	return ak.Role, true
}

// HasPermission checks if a role has the given permission.
func HasPermission(role Role, permission string) bool {
	perms, ok := RolePermissions[role]
	if !ok {
		return false
	}
	return perms[permission]
}

// Authorize checks if an API key has the given permission.
func (r *RBACManager) Authorize(key, permission string) bool {
	role, valid := r.ValidateKey(key)
	if !valid {
		return false
	}
	return HasPermission(role, permission)
}
